File: pi-extension/src/progress.ts
/**
 * progress.ts — Persistent task progress markers.
 *
 * Mirror of Python ``progress.py``.  Each task gets a stable id derived
 * from the user input text, persisted to ``<dir>/<task_id>.json``.
 * Re-running the same task returns the existing record (with completed
 * step ids) so the controller can resume.
 */

import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile, rename, appendFile, readdir } from "node:fs/promises";
import { existsSync } from "node:fs";
import { join } from "node:path";

export interface TaskProgress {
  taskId: string;
  userInput: string;
  created: number;
  updated: number;
  completedStepIds: string[];
  failedStepIds: string[];
  planSnapshot: Record<string, unknown>;
  checkpointData: Record<string, unknown>;
  status: "running" | "done" | "failed" | "abandoned";
}

/** Serialise a TaskProgress to the Python on-disk schema (snake_case keys).
 *  ``progress.py``'s ``TaskProgress`` dataclass + ``asdict()`` produces
 *  snake_case, and the README/header advertise that both runtimes can
 *  share a progress directory.  Cross-runtime resume requires us to
 *  match Python on the wire. */
function progressToPython(rec: TaskProgress): Record<string, unknown> {
  return {
    task_id: rec.taskId,
    user_input: rec.userInput,
    created: rec.created,
    updated: rec.updated,
    completed_step_ids: rec.completedStepIds,
    failed_step_ids: rec.failedStepIds,
    plan_snapshot: rec.planSnapshot,
    checkpoint_data: rec.checkpointData,
    status: rec.status,
  };
}

/** Parse from disk.  Accept the Python schema as authoritative AND the
 *  legacy camelCase shape that older extension builds wrote, so existing
 *  in-flight tasks still resume after the upgrade. */
function progressFromDisk(parsed: unknown): TaskProgress | undefined {
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) return undefined;
  const o = parsed as Record<string, unknown>;
  const pick = <T,>(snake: string, camel: string, fallback: T): T => {
    const v = o[snake] !== undefined ? o[snake] : o[camel];
    return (v ?? fallback) as T;
  };
  const taskId = pick("task_id", "taskId", "");
  if (!taskId) return undefined;
  const status = pick("status", "status", "running") as TaskProgress["status"];
  return {
    taskId,
    userInput: pick("user_input", "userInput", ""),
    created: pick("created", "created", Date.now() / 1000),
    updated: pick("updated", "updated", Date.now() / 1000),
    completedStepIds: pick("completed_step_ids", "completedStepIds", []),
    failedStepIds: pick("failed_step_ids", "failedStepIds", []),
    planSnapshot: pick("plan_snapshot", "planSnapshot", {}),
    checkpointData: pick("checkpoint_data", "checkpointData", {}),
    status,
  };
}

export class ProgressStore {
  private readonly dir: string;
  private readonly indexPath: string;

  constructor(directory: string) {
    this.dir = directory;
    this.indexPath = join(directory, "index.jsonl");
  }

  static deriveTaskId(userInput: string): string {
    const normalised = (userInput || "").split(/\s+/).filter(Boolean).join(" ").slice(0, 512);
    return createHash("sha1").update(normalised).digest("hex").slice(0, 12);
  }

  private async ensureDir(): Promise<void> {
    await mkdir(this.dir, { recursive: true });
  }

  private pathFor(taskId: string): string {
    return join(this.dir, `${taskId}.json`);
  }

  private async atomicWrite(path: string, data: TaskProgress): Promise<void> {
    const tmp = path + ".tmp";
    try {
      await writeFile(tmp, JSON.stringify(progressToPython(data), null, 2), "utf8");
      await rename(tmp, path);
    } catch {
      // best-effort write
    }
  }

  async load(taskId: string): Promise<TaskProgress | undefined> {
    await this.ensureDir();
    const p = this.pathFor(taskId);
    if (!existsSync(p)) return undefined;
    try {
      const text = await readFile(p, "utf8");
      return progressFromDisk(JSON.parse(text));
    } catch {
      return undefined;
    }
  }

  async openTask(userInput: string): Promise<TaskProgress> {
    await this.ensureDir();
    const taskId = ProgressStore.deriveTaskId(userInput);
    const existing = await this.load(taskId);
    if (existing && existing.status === "running") {
      existing.updated = Date.now() / 1000;
      return existing;
    }
    const now = Date.now() / 1000;
    const record: TaskProgress = {
      taskId,
      userInput,
      created: now,
      updated: now,
      completedStepIds: [],
      failedStepIds: [],
      planSnapshot: {},
      checkpointData: {},
      status: "running",
    };
    await this.atomicWrite(this.pathFor(taskId), record);
    // Index append is best-effort, mirroring atomicWrite() and the
    // Python ProgressStore.  Indexing is non-critical (resume uses
    // the per-task json files), so a transient FS error must not
    // abort the user's /autogui task.
    try {
      await appendFile(
        this.indexPath,
        JSON.stringify({
          task_id: record.taskId,
          user_input: userInput.slice(0, 160),
          updated: record.updated,
          status: record.status,
        }) + "\n",
        "utf8",
      );
    } catch {
      // swallow — per-task json file is the source of truth
    }
    return record;
  }

  async updatePlanSnapshot(rec: TaskProgress, snapshot: Record<string, unknown>): Promise<void> {
    rec.planSnapshot = { ...(snapshot ?? {}) };
    rec.updated = Date.now() / 1000;
    await this.atomicWrite(this.pathFor(rec.taskId), rec);
  }

  async markDone(rec: TaskProgress, stepId: string): Promise<void> {
    if (stepId && !rec.completedStepIds.includes(stepId)) rec.completedStepIds.push(stepId);
    rec.updated = Date.now() / 1000;
    await this.atomicWrite(this.pathFor(rec.taskId), rec);
  }

  async markFailed(rec: TaskProgress, stepId: string): Promise<void> {
    if (stepId && !rec.failedStepIds.includes(stepId)) rec.failedStepIds.push(stepId);
    rec.updated = Date.now() / 1000;
    await this.atomicWrite(this.pathFor(rec.taskId), rec);
  }

  async updateCheckpoint(rec: TaskProgress, data: Record<string, unknown>): Promise<void> {
    rec.checkpointData = { ...rec.checkpointData, ...(data ?? {}) };
    rec.updated = Date.now() / 1000;
    await this.atomicWrite(this.pathFor(rec.taskId), rec);
  }

  async finalize(rec: TaskProgress, status: TaskProgress["status"]): Promise<void> {
    rec.status = status;
    rec.updated = Date.now() / 1000;
    await this.atomicWrite(this.pathFor(rec.taskId), rec);
    // Match openTask()'s index-line shape (and the Python mirror) — the
    // index is the canonical "list of every task ever seen", so dropping
    // user_input here would make the schema inconsistent for any
    // consumer that relies on it for resume / listing.  Best-effort
    // append: finalize() runs at task end, so a successful run must
    // never surface as an extension error just because the index
    // sidecar couldn't be appended to.
    try {
      await appendFile(
        this.indexPath,
        JSON.stringify({
          task_id: rec.taskId,
          user_input: rec.userInput.slice(0, 160),
          updated: rec.updated,
          status,
        }) + "\n",
        "utf8",
      );
    } catch {
      // swallow — per-task json record is already on disk
    }
  }

  async listResumable(): Promise<TaskProgress[]> {
    await this.ensureDir();
    const entries: TaskProgress[] = [];
    try {
      const files = await readdir(this.dir);
      for (const f of files) {
        if (!f.endsWith(".json") || f === "index.jsonl") continue;
        try {
          const text = await readFile(join(this.dir, f), "utf8");
          const rec = progressFromDisk(JSON.parse(text));
          if (rec && rec.status === "running") entries.push(rec);
        } catch {
          // skip
        }
      }
    } catch {
      // dir missing — nothing to list
    }
    entries.sort((a, b) => b.updated - a.updated);
    return entries;
  }
}

File: progress.py
"""
progress.py — Persistent task progress markers.

Long, multi-step tasks should be resumable: a context reset, manual abort,
or process crash should not throw away the steps that already succeeded.
The progress store persists ``{task_id, plan, completed_step_ids,
checkpoint_data}`` to disk on every plan-step status change so the
controller can pick up exactly where it left off.

Format: one JSON file per task under ``logs/progress/<task_id>.json``.
The file is atomically replaced on each write to avoid half-written
state.  An ``index.jsonl`` lists every task ever seen so the user (or
the controller) can list resumable tasks.
"""

from __future__ import annotations

import hashlib
import json
import logging
import os
import time
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)


@dataclass
class TaskProgress:
    """In-memory mirror of a persisted task progress record."""

    task_id: str
    user_input: str
    created: float = field(default_factory=time.time)
    updated: float = field(default_factory=time.time)
    completed_step_ids: list[str] = field(default_factory=list)
    failed_step_ids: list[str] = field(default_factory=list)
    plan_snapshot: dict[str, Any] = field(default_factory=dict)
    checkpoint_data: dict[str, Any] = field(default_factory=dict)
    status: str = "running"   # running | done | failed | abandoned


class ProgressStore:
    """
    Disk-backed task progress.

    Methods are intentionally narrow: the controller calls
    ``open_task`` once per user request (it returns an existing record
    when the task_id matches an unfinished task), then ``mark_done`` /
    ``mark_failed`` as steps complete, and ``finalize`` when the whole
    task ends.
    """

    def __init__(self, directory: str = "logs/progress"):
        self._dir = Path(directory).expanduser()
        self._dir.mkdir(parents=True, exist_ok=True)
        self._index_path = self._dir / "index.jsonl"

    # ------------------------------------------------------------------
    # ID helper — derive a stable id from the task text so re-running the
    # same task picks up the previous record.
    # ------------------------------------------------------------------

    @staticmethod
    def derive_task_id(user_input: str) -> str:
        normalised = " ".join((user_input or "").split())[:512]
        return hashlib.sha1(normalised.encode("utf-8")).hexdigest()[:12]

    # ------------------------------------------------------------------
    # I/O
    # ------------------------------------------------------------------

    def _path_for(self, task_id: str) -> Path:
        return self._dir / f"{task_id}.json"

    def _atomic_write(self, path: Path, data: dict[str, Any]) -> None:
        tmp = path.with_suffix(path.suffix + ".tmp")
        try:
            tmp.write_text(
                json.dumps(data, ensure_ascii=False, indent=2),
                encoding="utf-8",
            )
            os.replace(tmp, path)
        except OSError as e:
            logger.warning("[progress] write failed for %s: %s", path, e)

    def _append_index(self, record: TaskProgress) -> None:
        line = {
            "task_id": record.task_id,
            "user_input": record.user_input[:160],
            "updated": record.updated,
            "status": record.status,
        }
        try:
            with self._index_path.open("a", encoding="utf-8") as f:
                f.write(json.dumps(line, ensure_ascii=False) + "\n")
        except OSError as e:
            logger.warning("[progress] index append failed: %s", e)

    def load(self, task_id: str) -> TaskProgress | None:
        p = self._path_for(task_id)
        if not p.exists():
            return None
        try:
            data = json.loads(p.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as e:
            logger.warning("[progress] could not parse %s: %s", p, e)
            return None
        try:
            return TaskProgress(**data)
        except TypeError as e:
            logger.warning("[progress] schema mismatch in %s: %s", p, e)
            return None

    # ------------------------------------------------------------------
    # Public interface
    # ------------------------------------------------------------------

    def open_task(self, user_input: str) -> TaskProgress:
        """
        Return an existing in-progress record for the same task text, or
        start a new one.  Finalised records (``done`` / ``failed`` /
        ``abandoned``) never resume — the user explicitly re-runs them.
        """
        task_id = self.derive_task_id(user_input)
        existing = self.load(task_id)
        if existing and existing.status == "running":
            existing.updated = time.time()
            return existing

        record = TaskProgress(task_id=task_id, user_input=user_input)
        self._atomic_write(self._path_for(task_id), asdict(record))
        self._append_index(record)
        return record

    def update_plan_snapshot(self, record: TaskProgress, snapshot: dict[str, Any]) -> None:
        record.plan_snapshot = dict(snapshot or {})
        record.updated = time.time()
        self._atomic_write(self._path_for(record.task_id), asdict(record))

    def mark_done(self, record: TaskProgress, step_id: str) -> None:
        if step_id and step_id not in record.completed_step_ids:
            record.completed_step_ids.append(step_id)
        record.updated = time.time()
        self._atomic_write(self._path_for(record.task_id), asdict(record))

    def mark_failed(self, record: TaskProgress, step_id: str) -> None:
        if step_id and step_id not in record.failed_step_ids:
            record.failed_step_ids.append(step_id)
        record.updated = time.time()
        self._atomic_write(self._path_for(record.task_id), asdict(record))

    def update_checkpoint(self, record: TaskProgress, data: dict[str, Any]) -> None:
        """Free-form checkpoint — used for "tab N of M done" style markers."""
        record.checkpoint_data.update(data or {})
        record.updated = time.time()
        self._atomic_write(self._path_for(record.task_id), asdict(record))

    def finalize(self, record: TaskProgress, *, status: str) -> None:
        record.status = status
        record.updated = time.time()
        self._atomic_write(self._path_for(record.task_id), asdict(record))
        self._append_index(record)

    def list_resumable(self) -> list[TaskProgress]:
        out: list[TaskProgress] = []
        for path in sorted(self._dir.glob("*.json")):
            if path.name == "index.jsonl":
                continue
            try:
                data = json.loads(path.read_text(encoding="utf-8"))
                rec = TaskProgress(**data)
            except Exception:  # pragma: no cover — best-effort listing
                continue
            if rec.status == "running":
                out.append(rec)
        out.sort(key=lambda r: r.updated, reverse=True)
        return out
